crates/bevy_ecs/src/schedule/graph/node.rs

```diff
@@ -1,42 +1,62 @@
-use core::fmt::Debug;
+use core::{fmt::Debug, hash::Hash};
 
-use crate::schedule::{SystemKey, SystemSetKey};
+use crate::schedule::graph::Direction;
 
-/// Unique identifier for a system or system set stored in a [`ScheduleGraph`].
+/// Types that can be used as node identifiers in a [`DiGraph`]/[`UnGraph`].
 ///
-/// [`ScheduleGraph`]: crate::schedule::ScheduleGraph
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub enum NodeId {
-    /// Identifier for a system.
-    System(SystemKey),
-    /// Identifier for a system set.
-    Set(SystemSetKey),
+/// [`DiGraph`]: crate::schedule::graph::DiGraph
+/// [`UnGraph`]: crate::schedule::graph::UnGraph
+pub trait GraphNodeId: Copy + Eq + Hash + Ord + Debug {
+    /// This [`GraphNodeId`] and a [`Direction`].
+    type Directed: DirectedGraphNodeId<Id = Self>;
+    /// Two of these [`GraphNodeId`]s.
+    type Pair: GraphNodeIdPair<Id = Self>;
 }
 
-impl NodeId {
-    /// Returns `true` if the identified node is a system.
-    pub const fn is_system(&self) -> bool {
-        matches!(self, NodeId::System(_))
+/// Types that are a [`GraphNodeId`] with a [`Direction`].
+pub trait DirectedGraphNodeId: Copy + Debug {
+    /// The type of [`GraphNodeId`] a [`Direction`] is paired with.
+    type Id: GraphNodeId;
+
+    /// Packs a [`GraphNodeId`] and a [`Direction`] into a single type.
+    fn new(id: Self::Id, direction: Direction) -> Self;
+
+    /// Unpacks a [`GraphNodeId`] and a [`Direction`] from this type.
+    fn unwrap(self) -> (Self::Id, Direction);
```

Comment:
`unwrap()` is usually used for panicking methods.  Would it make sense to call these `pack` and `unpack`?

Reply:
Agreed; this needs to be changed.

Reply:
Done.